new_platforms/scripts/VirtualMachineSgxSettings.psm1

```diff
@@ -0,0 +1,283 @@
+### -----------------------------------
+
+Class VirtualMachineSgx{
+    [bool] $IsSgxEnabled;
+    [UInt64] $SgxSize;
+    [String] $SgxLaunchControlDefault;
+    [UInt32] $SgxLaunchControlMode;
+
+    VirtualMachineSgx([System.Management.ManagementObject] $VmSgx) {
+        $This.IsSgxEnabled = $VmSgx.SgxEnabled;
+        $This.SgxSize = $VmSgx.SgxSize;
+        $This.SgxLaunchControlDefault = $VmSgx.SgxLaunchControlDefault;
+        $This.SgxLaunchControlMode = $VmSgx.SgxLaunchControlMode;
+    }
+}
+
+### -----------------------------------
+### Helper Functions
+### -----------------------------------
+
+Function Get-VMSgxMemorySettingData([Microsoft.HyperV.PowerShell.VirtualMachine] $Vm)
+{
+    $Mem =  Get-WmiObject -Namespace root/virtualization/v2 -Class Msvm_MemorySettingData |
+        Where-Object { $_.InstanceId -eq "Microsoft:$($Vm.Id)\$VM_MEMORY_SETTINGS_GUID" }
+
+    If (!$Mem)
+    {
+        Write-Error ($Strings.ERR_NO_MEM_SETTING_DATA_OBJECT -f $Vm.Name)
+    }
+
+    Return $Mem
+}
+
+Function Get-VMSgxManagementService()
+{
+    $Svc = Get-WmiObject -Namespace root/virtualization/v2 -Class Msvm_VirtualSystemManagementService |
+        Where-Object {
+            ($_.CreationClassName -eq "Msvm_VirtualSystemManagementService") -and
+            ($_.Name -eq "vmms") -and
+            ($_.SystemCreationClassName -eq "Msvm_ComputerSystem") -and
+            ($_.SystemName -eq $env:COMPUTERNAME)
+        }
+
+    If (!$Svc)
+    {
+        Write-Error $Strings.ERR_NO_VMM
+    }
+
+    return $Svc
+}
+
+### -----------------------------------
+### Get-VMSgx Cmdlet
+### -----------------------------------
+
+Function Get-VMSgx
+{
+    [CmdletBinding()]
+    Param
+    (
+        # Name of the VM to get the SGX settings for.
+        [Parameter(Mandatory=$True, ParameterSetName="ByName", Position=0)]
+        [ValidateNotNullOrEmpty()]
+        [String]$VmName,
+
+        # VM to get the SGX settings for.
+        [Parameter(Mandatory=$True, ParameterSetName="ByObject", Position=0)]
+        [ValidateNotNull()]
+        [Microsoft.HyperV.PowerShell.VirtualMachine]$Vm
+    )
+
+    Process
+    {
+        If (!$Vm)
+        {
+            $Vm = Get-VM $VmName
+        }
+
+        $Mem = Get-VMSgxMemorySettingData $Vm
+        Return [VirtualMachineSgx]::New($Mem)
+    }
+}
+
+### -----------------------------------
+### Set-VMSgx Cmdlet
+### -----------------------------------
+
+Function Set-VMSgx
+{
+    [CmdletBinding()]
+    Param
+    (
+        # Name of the VM to configure.
+        [Parameter(Mandatory=$True, ParameterSetName="ByName", Position=0)]
+        [ValidateNotNullOrEmpty()]
+        [String]$VmName,
+
+        # VM to configure.
+        [Parameter(Mandatory=$True, ParameterSetName="ByObject", Position=0)]
+        [ValidateNotNull()]
+        [Microsoft.HyperV.PowerShell.VirtualMachine]$Vm,
+
+        # Whether SGX is enabled in the VM.
+        [Parameter(Position=1)]
+        [Bool]$IsSgxEnabled = $False,
```

Comment:
Would it make sense to make this into a switch?

Reply:
I prefer to think of switches as the elements passed in through the command line and storage variables as boolean. Thoughts?

Reply:
Agreed, but I think I am missing something in this case. In this case, -IsSgxEnabled would be passed through command line or however the cmdlet is invoked. I would think about storage variable as a local function variable or a global variable. How are you distinguishing between storage boolean parameters and switches?

In my head, if the default was $true, then I would keep it as Bool. But if the default is $false, then it feels like a switch. Wanna enable it, add the switch.

Reply:
Speaking of which, if this is an input parameter, Is... does not sound to me like the right name.

Thinking more, I think I understand better your original intent. You want to have a way to specify both true and false, but not false implicitly by ignoring (even though that is what currently happens). Ok, I guess bool makes sense here, but would still consider renaming it, perhaps SgxEnabled?

Reply:
I've changed it to `SgxEnabled` everywhere.

The function checks against `$PSBoundParameters`, so `SgxEnabled` isn't read and nothing is done if you don't specify it.

Reply:
Given that you are checking against PSBoundParameters, does it make sense to have default values? Imho it does not, as they are not used. Seems confusing to me.

Reply:
Removed the defaults.